packages/pglite-sync/src/index.ts

```diff
@@ -113,123 +156,183 @@ async function createPlugin(
+            const remainingMessages: ChangeMessage<any>[] = []
+            let foundNonInsert = false
+            for (const message of messageAggregator) {
+              if (!foundNonInsert && message.headers.operation === 'insert') {
+                initialInserts.push(message as InsertChangeMessage)
+              } else {
+                foundNonInsert = true
+                remainingMessages.push(message)
+              }
+            }
+            if (initialInserts.length > 0) {
+              // As `COPY FROM` doesn't trigger a NOTIFY, we pop
+              // the last insert message and and add it to the be beginning
+              // of the remaining messages to be applied after the `COPY FROM`
+              remainingMessages.unshift(initialInserts.pop()!)
+            }
+            messageAggregator = remainingMessages
+
+            // Do the `COPY FROM` with initial inserts
+            if (initialInserts.length > 0) {
+              applyMessagesToTableWithCopy({
+                pg: tx,
+                table: options.table,
+                schema: options.schema,
+                messages: initialInserts as InsertChangeMessage[],
+                mapColumns: options.mapColumns,
+                primaryKey: options.primaryKey,
+                debug,
+              })
+              // We don't want to do a `COPY FROM` again after that
+              doCopy = false
+            }
+          }
+
+          for (const changeMessage of messageAggregator) {
+            await applyMessageToTable({
+              pg: tx,
+              table: options.table,
+              schema: options.schema,
+              message: changeMessage,
+              mapColumns: options.mapColumns,
+              primaryKey: options.primaryKey,
+              debug,
+            })
+          }
+
+          if (
+            options.shapeKey &&
+            messageAggregator.length > 0 &&
+            stream.shapeHandle !== undefined
+          ) {
+            await updateShapeSubscriptionState({
+              pg: tx,
+              metadataSchema,
+              shapeKey: options.shapeKey,
+              shapeId: stream.shapeHandle,
+              lastOffset:
+                messageAggregator[messageAggregator.length - 1].offset,
+            })
+          }
+        })
+        if (debug) console.timeEnd('commit')
+        messageAggregator = []
+        // Await a timeout to start a new task and  allow other connections to do work
+        await new Promise((resolve) => setTimeout(resolve, 0))
+      }
+
+      const throttledCommit = async () => {
+        const now = Date.now()
+        if (options.commitThrottle && debug)
+          console.log(
+            'throttled commit: now:',
+            now,
+            'lastCommitAt:',
+            lastCommitAt,
+            'diff:',
+            now - lastCommitAt,
+          )
+        if (
+          options.commitThrottle &&
+          now - lastCommitAt < options.commitThrottle
```

Comment:
I don't like that this throttler might not run the last received commit - perhaps add a utility with a more full fledged throttler or use a utility like lodash's? Ideally the last commit call should always execute after the throttle period and not have to wait for one more if it was not run due to throttling.

edit: I realise that the behaviour I describe would introduce complexity in awaiting things, handling errors, etc - perhaps something like having a unified throttle timer mechanism that checks if there are messages to commit at the end of every throttle timeout and inserts a "pseudo-batch" of messages to trigger processing might work - let me know if the "last messages not being committed" issue is not really an issue and I'm missing something

Reply:
My thinking was that as we will always end with an `up-to-date` in the protocol, and that forces an un-throttled commit, that we may not need to have a timeout like a "normal" throttling utility?